main.md

```diff
@@ -0,0 +1,465 @@
+            for boxes, labels in zip(batch["boxes"], batch["labels"])
+        ]
+        loss_dict = model(inputs, targets)
+        losses = sum(loss for loss in loss_dict.values())
+
+        optimizer.zero_grad()
+        losses.backward()
+        optimizer.step()
+
+        if lr_scheduler is not None:
+            lr_scheduler.step()
+
+        session.report(
+            {
+                "losses": losses.item(),
+                "epoch": epoch,
+                "lr": optimizer.param_groups[0]["lr"],
+                **{key: value.item() for key, value in loss_dict.items()},
+            }
+        )
+
+
+def train_loop_per_worker(config):
+    model = models.detection.fasterrcnn_resnet50_fpn(num_classes=21)
+    model = ray.train.prepare_model(model)
+    parameters = [p for p in model.parameters() if p.requires_grad]
+    optimizer = torch.optim.SGD(
+        parameters,
+        lr=config["lr"],
+        momentum=config["momentum"],
+        weight_decay=config["weight_decay"],
+    )
+    lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(
+        optimizer, milestones=config["lr_steps"], gamma=config["lr_gamma"]
+    )
+
+    for epoch in range(0, config["epochs"]):
+        train_one_epoch(
+            model=model,
+            optimizer=optimizer,
+            batch_size=config["batch_size"],
+            epoch=epoch,
+        )
+        lr_scheduler.step()
+        checkpoint = Checkpoint.from_dict(
+            {
+                "model": model.module.state_dict(),
+                "optimizer": optimizer.state_dict(),
+                "lr_scheduler": lr_scheduler.state_dict(),
+                "config": config,
+                "epoch": epoch,
+            }
+        )
+        session.report({}, checkpoint=checkpoint)
+```
+
+### Fine-tune the model
+
+
+Once you've defined the training loop, create a `TorchTrainer` and pass the training
+loop to the constructor. Then, call `TorchTrainer.fit` to train the model.
+
+```python
+from ray.air.config import ScalingConfig
+from ray.train.torch import TorchTrainer
+
+
+trainer = TorchTrainer(
+    train_loop_per_worker=train_loop_per_worker,
+    train_loop_config={
+        "batch_size": 2,
+        "lr": 0.02,
+        "epochs": 26,
+        "momentum": 0.9,
+        "weight_decay": 1e-4,
+        "lr_steps": [16, 22],
+        "lr_gamma": 0.1,
+    },
+    scaling_config=ScalingConfig(num_workers=8, use_gpu=True),
+    datasets={"train": train_dataset},
+)
+results = trainer.fit()
+```
+
+## Evaluate the model on test data
+
+
+### Define a custom predictor
+
+
+`Predictors` perform inference on batches of data.
+
+To make `fasterrcnn_resnet50_fpn` outputs compatible with the `Predictor` interface,
+subclass  `TorchPredictor` and override the `_predict_numpy` method.
+
+```python
+import collections
+
+from ray.train.torch import TorchPredictor
+from ray.air.util.tensor_extensions.pandas import _create_possibly_ragged_ndarray
+
+
+class CustomTorchPredictor(TorchPredictor):
```

Comment:
Just to make sure, we’re planning to upstream this to ray?

Reply:
Yeah, created an issue to track: https://github.com/ray-project/ray/issues/32064.